accord-core/src/main/java/accord/coordinate/Coordinate.java

```diff
@@ -1,46 +1,283 @@
+    {
+        public ShardTracker(Shard shard)
+        {
+            super(shard);
+        }
+
+        @Override
+        public boolean includeInFastPath(Node.Id node, boolean withFastPathTimestamp)
+        {
+            return withFastPathTimestamp && shard.fastPathElectorate.contains(node);
+        }
+
+        @Override
+        public boolean hasMetFastPathCriteria()
+        {
+            return fastPathAccepts >= shard.fastPathQuorumSize;
+        }
+    }
+
+    static class PreacceptTracker extends FastPathTracker<ShardTracker>
+    {
+        volatile long supersedingEpoch = -1;
+        private final boolean fastPathPermitted;
+        private final Set<Id> successes = new HashSet<>();
+        private Set<Id> failures;
+
+        public PreacceptTracker(Topologies topologies, boolean fastPathPermitted)
+        {
+            super(topologies, Coordinate.ShardTracker[]::new, Coordinate.ShardTracker::new);
+            this.fastPathPermitted = fastPathPermitted;
+        }
+
+        public PreacceptTracker(Topologies topologies)
+        {
+            this(topologies, topologies.fastPathPermitted());
+        }
+
+        @Override
+        public boolean failure(Id node)
+        {
+            if (failures == null)
+                failures = new HashSet<>();
+            failures.add(node);
+            return super.failure(node);
+        }
+
+        @Override
+        public void recordSuccess(Id node, boolean withFastPathTimestamp)
+        {
+            successes.add(node);
+            super.recordSuccess(node, withFastPathTimestamp);
+        }
+
+        public void recordSuccess(Id node)
+        {
+            recordSuccess(node, false);
+        }
+
+        public synchronized boolean recordSupersedingEpoch(long epoch)
+        {
+            if (epoch <= supersedingEpoch)
+                return false;
+            supersedingEpoch = epoch;
+            return true;
+        }
+
+        public boolean hasSupersedingEpoch()
+        {
+            return supersedingEpoch > 0;
+        }
+
+        public PreacceptTracker withUpdatedTopologies(Topologies topologies)
+        {
+            PreacceptTracker tracker = new PreacceptTracker(topologies, false);
+            successes.forEach(tracker::recordSuccess);
+            if (failures != null)
+                failures.forEach(tracker::failure);
+            return tracker;
+        }
+
+        @Override
+        public boolean hasMetFastPathCriteria()
+        {
+            return fastPathPermitted && super.hasMetFastPathCriteria();
+        }
+
+        boolean shouldSlowPathAccept()
+        {
+            return (!fastPathPermitted || !hasInFlight()) && hasReachedQuorum();
+        }
+    }
+
+    final Node node;
+    final TxnId txnId;
+    final Txn txn;
+    final Key homeKey;
+
+    private PreacceptTracker tracker;
+    private final List<PreAcceptOk> preAcceptOks = new ArrayList<>();
+    private boolean isPreAccepted;
```

Comment:
If we decide to not separate preaccept from coordinate: `isDone` might be too generic for the Coordinate class, but `isPreaccepted` implies (to me) that preaccept was successful, and is odd to see set to true in `onFailure`. Maybe `preacceptDone` or something similar?